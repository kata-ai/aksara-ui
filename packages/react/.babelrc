{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "browsers": ["last 2 chrome versions", "last 2 firefox versions"]
        }
      }
    ],
    "@babel/preset-react",
    "@babel/preset-typescript"
  ],
  "plugins": [
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "babel-plugin-styled-components"
  ],
  "env": {
    "test": {
      "plugins": ["@babel/plugin-transform-modules-commonjs"]
    }
  }
}
