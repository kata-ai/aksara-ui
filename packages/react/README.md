# @aksara-ui/react

> React implementation of Aksara Design System, from Kata.ai.

This package is part of the Aksara monorepo. Please read the README in [the Aksara repository](https://github.com/kata-ai/aksara-ui) for detailed documentation about this package.
