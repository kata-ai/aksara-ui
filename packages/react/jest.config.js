module.exports = {
  collectCoverageFrom: [
    'src/**/*.{ts,tsx}',
    '!src/**/*.stories.{ts,tsx}',
    '!src/**/*.story.{ts,tsx}',
    '!src/**/index.{ts,tsx}',
    '!src/stories/**/*.{ts,tsx}',
    '!src/utils/storybook/**/*.{ts,tsx}',
    '!src/index.{ts,tsx}',
  ],
  coverageReporters: ['json', 'lcov'],
  moduleDirectories: ['node_modules', '<rootDir>/src'],
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
  modulePathIgnorePatterns: ['lib', 'lib-es6'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTests.ts'],
  testPathIgnorePatterns: ['<rootDir>/dist/'],
  transform: {
    '^.+\\.(js|jsx|ts|tsx|mjs)$': 'babel-jest',
    '^.+\\.css$': '<rootDir>/config/jest/cssTransform.js',
    '^(?!.*\\.(js|jsx|ts|tsx|mjs|css|json)$)': '<rootDir>/config/jest/fileTransform.js',
  },
};
