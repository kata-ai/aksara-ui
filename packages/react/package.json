{
  "name": "@aksara-ui/react",
  "version": "1.0.1-canary.4",
  "description": "React implementation of Aksara Design System, from Kata.ai.",
  "main": "dist/index.js",
  "module": "dist/core.esm.js",
  "typings": "dist/index.d.ts",
  "aksara:src": "./src/index.ts",
  "repository": "https://github.com/kata-ai/wicara.git",
  "author": "Kata.ai <dev@kata.ai>",
  "license": "Apache-2.0",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsdx build --transpileOnly --tsconfig tsconfig.build.json",
    "build-storybook": "build-storybook --quiet",
    "clean": "rimraf dist",
    "format": "prettier --write \"src/**/*.{js,jsx,ts,tsx}\"",
    "lint": "eslint \"src/**/*.{js,jsx,ts,tsx}\"",
    "prepublish": "yarn build",
    "report-coverage": "codecov -f ./coverage/*.json",
    "start": "tsdx watch --tsconfig tsconfig.build.json --verbose",
    "storybook": "start-storybook -p 9001",
    "storybook:build": "build-storybook --quiet",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "type-check": "tsc --emitDeclarationOnly"
  },
  "dependencies": {
    "@aksara-ui/core": "^1.0.1-canary.1",
    "@popperjs/core": "^2.10.2",
    "@reach/accordion": "^0.16.1",
    "@reach/visually-hidden": "^0.16.0",
    "@styled-system/css": "^5.1.5",
    "@styled-system/should-forward-prop": "^5.1.5",
    "@styled-system/theme-get": "^5.1.2",
    "@types/styled-system": "^5.1.13",
    "@types/styled-system__css": "^5.0.16",
    "@types/styled-system__theme-get": "^5.0.1",
    "clsx": "^1.1.1",
    "csstype": "^3.0.9",
    "deepmerge": "^4.2.2",
    "focus-trap": "^6.6.0",
    "polished": "^4.1.3",
    "react-fast-compare": "^3.2.0",
    "react-popper": "^2.2.5",
    "react-popper-tooltip": "^4.3.0",
    "react-transition-group": "^4.4.2",
    "styled-system": "^5.1.4",
    "tslib": "^2.3.1"
  },
  "peerDependencies": {
    "@aksara-ui/icons": "^1.0.0-canary.0",
    "react": "^16.10.0 || ^17.0.0",
    "react-dom": "^16.10.0 || ^17.0.0",
    "styled-components": "^5.1.0"
  },
  "devDependencies": {
    "@aksara-ui/icons": "^1.0.0",
    "@babel/core": "^7.15.0",
    "@storybook/addon-actions": "^6.3.7",
    "@storybook/addon-essentials": "^6.3.7",
    "@storybook/addon-knobs": "^6.3.0",
    "@storybook/addon-links": "^6.3.7",
    "@storybook/react": "^6.3.7",
    "@storybook/theming": "^6.3.7",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.0.0",
    "@types/jest": "^26.0.24",
    "@types/react-transition-group": "^4.4.2",
    "@types/styled-components": "^5.1.12",
    "babel-loader": "^8.2.2",
    "babel-plugin-styled-components": "^1.13.2",
    "babel-preset-react-app": "^10.0.0",
    "codecov": "^3.8.3",
    "copee": "^1.0.6",
    "jest": "^26.6.3",
    "jest-styled-components": "^7.0.5",
    "react": "^17.0.2",
    "react-docgen-typescript-loader": "^3.6.0",
    "react-dom": "^17.0.2",
    "styled-components": "^5.3.0"
  }
}
