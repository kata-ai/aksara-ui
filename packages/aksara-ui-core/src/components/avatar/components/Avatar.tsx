import * as React from 'react';
import styled from 'styled-components';
import { IconUser, IconAudience } from '@aksara-ui/icons';

import { AvatarBase, AvatarBaseProps } from '../styles';
import { Text } from '../../../foundations';
import getInitials from '../utils/getInitials';

export interface AvatarProps extends AvatarBaseProps {
  /** The avatar's image source. */
  src?: any;
  /** Additional CSS classes to give to the component. */
  className?: string;
  /** Alt text for the avatar */
  alt?: string;
  /** Additional CSS styles to give to the component. */
  style?: React.CSSProperties;
}

const Root = styled('div')`
  ${AvatarBase};
`;

/** Resizable avatar component. */
const Avatar: React.FC<AvatarProps> = ({ src, size, shape, className, style, alt, name, type, ...rest }) => {
  const renderAvatarIcon = () => {
    return type === 'user' ? <IconUser fill="currentColor" /> : <IconAudience fill="currentColor" />;
  };

  const renderInitials = () => {
    if (size === 24 || src || !name) {
      return null;
    }

    return <Text scale={300}>{getInitials(name)}</Text>;
  };

  const renderImage = () => {
    if (src) {
      return <img src={src} alt={alt} />;
    }

    if (!name) {
      return renderAvatarIcon();
    }

    return null;
  };

  return (
    <Root className={className} style={style} size={size} shape={shape} {...rest}>
      {renderInitials()}
      {renderImage()}
    </Root>
  );
};

Avatar.defaultProps = {
  size: 40,
  alt: '',
  color: 'indigo',
  type: 'user',
  shape: 'rounded',
};

Avatar.displayName = 'Avatar';

export default Avatar;
