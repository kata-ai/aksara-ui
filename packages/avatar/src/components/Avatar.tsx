import * as React from 'react';
import styled from 'styled-components';

import { AvatarBase } from '../styles';
import themes from '../theme';

export interface AvatarProps {
  src: any;
  /** Size of the avatar. */
  size?: number;
  /** Additional CSS classes to give to the component. */
  className?: string;
  /** Additional CSS styles to give to the component. */
  style?: React.CSSProperties;
}

export default class Avatar extends React.Component<AvatarProps> {
  static defaultProps = {
    size: 40
  };

  render() {
    const { src, size, className, style } = this.props;

    return (
      <Root
        className={className}
        style={style}
        theme={themes['defaultTheme']}
        size={size}
      >
        <Image src={src} />
      </Root>
    );
  }
}

const Image = styled('img')``;

const Root = styled('div')`
  ${AvatarBase};
`;
