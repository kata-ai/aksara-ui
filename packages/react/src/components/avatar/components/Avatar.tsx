import * as React from 'react';
import { avatarSizeOptions } from '../../../theme/component-styles/avatar';
import { VisuallyHidden } from '../../../helpers';
import { Box, BoxProps } from '../../../layout';
import { useComponentStyles } from '../../../system';
import { Text } from '../../../typography';
import getInitials from '../utils/getInitials';
import Presence, { PresenceProps } from './Presence';

type AvatarSize = 'sm' | 'md' | 'lg' | 'xl';
const bgColors = ['indigo06', 'red07', 'yellow07', 'blue07', 'green07'];
export interface AvatarProps extends Omit<BoxProps, 'size'> {
  /** Additional CSS classes to give to the component. */
  className?: string;
  /** Additional CSS styles to give to the component. */
  style?: React.CSSProperties;
  /** Name for initials */
  name?: string;
  /** The avatar's image source. */
  src?: any;
  /** Alt text for the avatar. */
  alt?: string;
  /** The background color of the avatar. */
  bg?: string;
  /** The icon that renders with the avatar. */
  icon?: React.ComponentType<any>;
  /** Size of the avatar. */
  size?: AvatarSize;
  /** Presence option */
  presence?: PresenceProps;
}

function iconSizes(size: AvatarSize = 'lg') {
  return avatarSizeOptions[size] - 8;
}

const initialSizes: Record<AvatarSize, string> = {
  sm: '8px',
  md: '12px',
  lg: '14px',
  xl: '20px',
};

/** Resizable avatar component. */
const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>(
  (
    { className, style, src, alt = undefined, name, size = 'lg', color = 'white', bg, icon, presence, ...rest },
    ref
  ) => {
    const styles = useComponentStyles('avatar', { size: typeof size === 'string' ? size : undefined });

    const renderInitials = () => {
      return (
        <Text fontSize={initialSizes[size]} lineHeight={1} role="presentation">
          {name ? getInitials(name) : '??'}
        </Text>
      );
    };

    const renderImage = () => {
      if (src) {
        return (
          <Box
            as="img"
            sx={{ position: 'absolute', width: '100%', height: '100%', objectFit: 'cover' }}
            src={src}
            alt={name || alt}
          />
        );
      }

      if (icon) {
        return React.createElement(icon, { fill: 'currentColor', size: iconSizes(size), 'aria-hidden': true });
      }

      return renderInitials();
    };

    const renderPresence = (option: PresenceProps) => {
      return <Presence {...option} size={size} />;
    };

    const bgColorsRandomizer = () => {
      if (!name) {
        return 'indigo06';
      }
      const names = name.split(' ');
      let totalValue = 0;
      names.forEach(item => {
        const utfValue = new TextEncoder().encode(`${item[0]}${item[item.length - 1]}`);
        totalValue += utfValue[0] + utfValue[1];
      });
      const indexColor = totalValue % bgColors.length;
      return bgColors[indexColor];
    };

    const renderAvatar = () => {
      return (
        <Box
          data-testid="avatar-container"
          as="span"
          ref={ref}
          className={className}
          style={style}
          color={color}
          bg={bg || bgColorsRandomizer()}
          sx={{ ...styles }}
          width={typeof size === 'number' ? size : undefined}
          height={typeof size === 'number' ? size : undefined}
          borderRadius={9999}
          {...rest}
        >
          {name && <VisuallyHidden>{name}</VisuallyHidden>}
          {renderImage()}
        </Box>
      );
    };
    if (presence) {
      return (
        <Box position="relative" data-testid="avatar-with-presence">
          {renderAvatar()}
          {renderPresence(presence)}
        </Box>
      );
    }
    return renderAvatar();
  }
);

Avatar.displayName = 'Avatar';

export default Avatar;
