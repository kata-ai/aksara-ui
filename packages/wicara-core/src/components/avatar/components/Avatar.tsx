import * as React from 'react';
import styled from 'styled-components';

import { AvatarBase } from '../styles';

export interface AvatarProps {
  /** The avatar's image source. */
  src: any;
  /** Size of the avatar. */
  size?: number;
  /** Additional CSS classes to give to the component. */
  className?: string;
  /** Alt text for the avatar */
  alt?: string;
  /** Additional CSS styles to give to the component. */
  style?: React.CSSProperties;
}

const Root = styled('div')`
  ${AvatarBase};
`;

/** Resizable avatar component. */
const Avatar: React.FC<AvatarProps> = ({ src, size, className, style, alt }) => {
  return (
    <Root className={className} style={style} size={size}>
      <img src={src} alt={alt} />
    </Root>
  );
};

Avatar.defaultProps = {
  size: 40,
  alt: ''
};

Avatar.displayName = 'Avatar';

export default Avatar;
