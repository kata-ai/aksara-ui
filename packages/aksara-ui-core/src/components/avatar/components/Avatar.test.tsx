import * as React from 'react';
import { render } from '@testing-library/react';

import Avatar from './Avatar';

describe('components/Avatar', () => {
  describe('<Avatar />', () => {
    test('renders on large size by default', () => {
      const { container } = render(<Avatar src="" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders in correct sizes', () => {
      const { container } = render(<Avatar size={24} src="" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders initials correctly', () => {
      const { getByText } = render(<Avatar name="Adry Muhammad" />);

      expect(getByText('AM')).toBeInTheDocument();
    });
  });
});
