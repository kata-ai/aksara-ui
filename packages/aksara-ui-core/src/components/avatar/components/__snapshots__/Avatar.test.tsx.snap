// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Avatar <Avatar /> renders in correct sizes 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 24px;
  height: 24px;
  border: 1px solid #fff;
  border-radius: 24px;
  background-color: indigo03;
  color: white;
  font-size: 1rem;
  line-height: 1;
  vertical-align: middle;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 24px;
  height: 24px;
}

.c0 img {
  position: absolute;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

<div
  class="c0"
  color="indigo"
  shape="rounded"
  size="24"
>
  <svg
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M22 13a10 10 0 10-16.466 7.62c.01-.062.038-.187.038-.187C6.018 18.493 8.722 17 12 17s5.982 1.493 6.428 3.433c0 0 .027.125.038.188A9.972 9.972 0 0022 13zm-6.141-2l-.41 2a3.484 3.484 0 01-6.899 0l-.409-2a4 4 0 117.718 0z"
      fill="currentColor"
    />
  </svg>
</div>
`;

exports[`components/Avatar <Avatar /> renders on large size by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 40px;
  height: 40px;
  border: 1px solid #fff;
  border-radius: 40px;
  background-color: indigo03;
  color: white;
  font-size: 1rem;
  line-height: 1;
  vertical-align: middle;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 40px;
  height: 40px;
}

.c0 img {
  position: absolute;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

<div
  class="c0"
  color="indigo"
  shape="rounded"
  size="40"
>
  <svg
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M22 13a10 10 0 10-16.466 7.62c.01-.062.038-.187.038-.187C6.018 18.493 8.722 17 12 17s5.982 1.493 6.428 3.433c0 0 .027.125.038.188A9.972 9.972 0 0022 13zm-6.141-2l-.41 2a3.484 3.484 0 01-6.899 0l-.409-2a4 4 0 117.718 0z"
      fill="currentColor"
    />
  </svg>
</div>
`;
